core/lib/dal/src/consensus_dal.rs

```diff
@@ -47,6 +54,158 @@ impl ConsensusDal<'_, '_> {
+    }
+
+    /// Fetches the last consensus certificate.
+    /// Currently certificates are NOT generated synchronously with miniblocks,
+    /// so it might NOT be the certificate for the last miniblock.
+    pub async fn last_certificate(&mut self) -> anyhow::Result<Option<validator::CommitQC>> {
+        let Some(row) = sqlx::query!(
+            r#"
+            SELECT
+                certificate
+            FROM
+                miniblocks_consensus
+            ORDER BY
+                number DESC
+            LIMIT
+                1
+            "#
+        )
+        .fetch_optional(self.storage.conn())
+        .await?
+        else {
+            return Ok(None);
+        };
+        Ok(Some(zksync_protobuf::serde::deserialize(row.certificate)?))
+    }
+
+    /// Fetches the consensus certificate for the miniblock with the given `block_number`.
+    pub async fn certificate(
+        &mut self,
+        block_number: validator::BlockNumber,
+    ) -> anyhow::Result<Option<validator::CommitQC>> {
+        let Some(row) = sqlx::query!(
+            r#"
+            SELECT
+                certificate
+            FROM
+                miniblocks_consensus
+            WHERE
+                number = $1
+            "#,
+            i64::try_from(block_number.0)?
+        )
+        .fetch_optional(self.storage.conn())
+        .await?
+        else {
+            return Ok(None);
+        };
+        Ok(Some(zksync_protobuf::serde::deserialize(row.certificate)?))
+    }
+
+    /// Converts the miniblock `block_number` into consensus payload. `Payload` is an
+    /// opaque format for the miniblock that consensus understands and generates a
+    /// certificate for it.
+    pub async fn block_payload(
+        &mut self,
+        block_number: validator::BlockNumber,
+        operator_address: Address,
+    ) -> anyhow::Result<Option<Payload>> {
+        let block_number = MiniblockNumber(block_number.0.try_into()?);
+        let Some(block) = self
+            .storage
+            .sync_dal()
+            .sync_block_inner(block_number)
+            .await?
+        else {
+            return Ok(None);
+        };
+        let transactions = self
+            .storage
+            .sync_dal()
+            .sync_block_transactions(block_number)
+            .await?;
+        Ok(Some(block.into_payload(operator_address, transactions)))
+    }
+
+    /// Inserts a certificate for the miniblock `cert.header().number`.
+    /// It verifies that
+    /// * the certified payload matches the miniblock in storage
+    /// * the `cert.header().parent` matches the parent miniblock.
+    /// * the parent block already has a certificate.
+    /// NOTE: This is an extra secure way of storing a certificate,
+    /// which will help us to detect bugs in the consensus implementation
+    /// while it is "fresh". If it turns out to take too long,
+    /// we can remove the verification checks later.
```

Comment:
@pompon0 ideally `dal` code should only serve as the DB interface and not do any business-logic. Do you think there is a better place for such validation? I'm not concerned about performance by the way - I like having those checks.

Reply:
these are storage consistent checks. I would like to avoid exposing public dal methods which allow to corrupt invariants that are checked here.

Reply:
@pompon0 I understand the reasoning but 

> avoid exposing public dal methods which allow to corrupt invariants

This is not how we approached the DAL abstraction in the first place - it's a dump wrapper around sql queries. We also try to enforce invariants on the schema design level (e.g. via normalization) - so that it's impossible to represent an invalid state. 

Having said that, I won't insist - and I know in other places we weren't that strict. Feel free to leave as is if you like - for me it was only important to let you know the reasoning